File: api/service/api/social.go
// service/api/social.go

package api

import (
	"fmt"

	"github.com/gin-gonic/gin"
	"github.com/jcytp/kifup-api/common/handler"
	"github.com/jcytp/kifup-api/service/dao"
	"github.com/jcytp/kifup-api/service/model"
)

func LikeKifu(c *gin.Context) (string, error) {
	accountID := handler.GetActorID(c)
	kifuID := c.GetString("kifuID")

	// 公開棋譜の存在確認
	kifu, err := dao.GetKifu(kifuID)
	if err != nil {
		return "Failed to get kifu", err
	}
	if !kifu.IsPublic { // 公開棋譜のみ
		return "Access denied", fmt.Errorf("unauthorized access to private kifu")
	}

	// いいねを追加してカウンターを更新
	if err := dao.InsertKifuLike(kifuID, accountID); err != nil {
		return "Failed to insert like", err
	}
	if err := dao.IncrementKifuLikeCount(kifuID); err != nil {
		return "Failed to update like count", err
	}

	return "", nil
}

func UnlikeKifu(c *gin.Context) (string, error) {
	accountID := handler.GetActorID(c)
	kifuID := c.GetString("kifuID")

	// いいねを削除してカウンターを更新
	if err := dao.DeleteKifuLike(kifuID, accountID); err != nil {
		return "Failed to delete kifu like", err
	}
	if err := dao.DecrementKifuLikeCount(kifuID); err != nil {
		return "Failed to update kifu like count", err
	}

	return "", nil
}

type requestPostKifuComment struct {
	Content string `json:"content" binding:"required,min=1,max=1000"`
}

func PostKifuComment(c *gin.Context, req requestPostKifuComment) (string, error) {
	accountID := handler.GetActorID(c)
	kifuID := c.GetString("kifuID")

	// 棋譜の存在確認
	kifu, err := dao.GetKifu(kifuID)
	if err != nil {
		return "Failed to get kifu", err
	}
	if !kifu.IsPublic { // 公開棋譜のみ
		return "Access denied", fmt.Errorf("unauthorized access to private kifu")
	}

	// コメントを追加してカウンターを更新
	comment := &model.KifuComment{
		KifuID:    kifuID,
		AccountID: accountID,
		Content:   req.Content,
	}
	commentID, err := dao.InsertKifuComment(comment)
	if err != nil {
		return "Failed to insert kifu comment", err
	}
	if err := dao.IncrementKifuCommentCount(kifuID); err != nil {
		return "Failed to update kifu comment count", err
	}

	return commentID, nil
}

func ListKifuComments(c *gin.Context) (*[]*model.KifuCommentResponse, string, error) {
	accountID := handler.GetActorID(c)
	kifuID := c.GetString("kifuID")

	// 棋譜の存在確認
	kifu, err := dao.GetKifu(kifuID)
	if err != nil {
		return nil, "Failed to get kifu", err
	}
	if !kifu.IsPublic && kifu.AccountID != accountID { // 非公開の場合は所有者のみ
		return nil, "Access denied", fmt.Errorf("unauthorized access to private kifu")
	}

	// コメント一覧を取得
	comments, err := dao.ListKifuComments(kifuID)
	if err != nil {
		return nil, "Failed to get comments", err
	}

	// レスポンスを構築（アカウント情報も含める）
	responses := make([]*model.KifuCommentResponse, 0, len(comments))
	for _, comment := range comments {
		account, err := dao.GetAccountByID(comment.AccountID)
		if err != nil {
			return nil, "Failed to get account info", err
		}
		responses = append(responses, comment.ToResponse(account))
	}

	return &responses, "", nil
}

File: frontend/src/lib/apis/social.ts
// src/lib/apis/social.ts

import { API, type ApiResult } from '$lib/types/API';

export const addKifuLike = async (kifuId: string): Promise<ApiResult> => {
  const result = await API.post(`/api/kifu/${kifuId}/like`, null, true);
  if (!result.ok) {
    console.error('add kifu like error');
    result.data = 'いいねの追加に失敗しました。';
  }
  return result;
};

export const removeKifuLike = async (kifuId: string): Promise<ApiResult> => {
  const result = await API.delete(`/api/kifu/${kifuId}/like`, null, true);
  if (!result.ok) {
    console.error('remove kifu like error');
    result.data = 'いいねの削除に失敗しました。';
  }
  return result;
};

export const getKifuComments = async (kifuId: string): Promise<ApiResult> => {
  const result = await API.get(`/api/kifu/${kifuId}/comments`, null, false);
  if (!result.data) {
    console.error('get kifu comments error: no data');
    result.ok = false;
    result.data = 'コメントの取得に失敗しました。';
  }
  return result;
};

export const addKifuComment = async (kifuId: string, content: string): Promise<ApiResult> => {
  const params = { content };
  const result = await API.post(`/api/kifu/${kifuId}/comment`, params, true);
  if (!result.ok) {
    console.error('add kifu comment error');
    result.data = 'コメントの投稿に失敗しました。';
  }
  return result;
};
